internal/impl/mongodb/input.go

```diff
@@ -135,33 +176,111 @@ func (m *mongoInput) Connect(ctx context.Context) error {
 	return nil
 }
 
-func (m *mongoInput) Read(ctx context.Context) (*service.Message, service.AckFunc, error) {
+func (m *mongoInput) ReadBatch(ctx context.Context) (service.MessageBatch, service.AckFunc, error) {
+	i := 0
+	batch := make(service.MessageBatch, m.batchSize)
+
 	if m.cursor == nil {
 		return nil, nil, service.ErrNotConnected
 	}
-	if !m.cursor.Next(ctx) {
-		return nil, nil, service.ErrEndOfInput
-	}
-	var decoded any
-	if err := m.cursor.Decode(&decoded); err != nil {
-		return nil, nil, err
-	}
 
-	data, err := bson.MarshalExtJSON(decoded, m.marshalCanon, false)
-	if err != nil {
-		return nil, nil, err
-	}
+	for m.cursor.Next(ctx) {
+		msg := service.NewMessage(nil)
+		msg.MetaSet("mongo_database", m.database.Name())
+		msg.MetaSet("mongo_collection", m.collection)
 
-	msg := service.NewMessage(nil)
-	msg.SetBytes(data)
-	return msg, func(ctx context.Context, err error) error {
-		return nil
-	}, nil
+		var decoded any
+		if err := m.cursor.Decode(&decoded); err != nil {
+			msg.SetError(err)
+		} else {
+			data, err := bson.MarshalExtJSON(decoded, m.marshalCanon, false)
+			if err != nil {
+				msg.SetError(err)
+			}
+			msg.SetBytes(data)
+		}
+		batch[i] = msg
+		i++
+		m.count++
+
+		if m.cursor.RemainingBatchLength() == 0 {
+			return batch[:i], func(ctx context.Context, err error) error {
+				return nil
+			}, nil
+		}
+	}
+	return nil, nil, service.ErrEndOfInput
 }
 
 func (m *mongoInput) Close(ctx context.Context) error {
 	if m.cursor != nil && m.client != nil {
+		m.logger.Infof("Got %d documents from '%s' collection", m.count, m.collection)
```

Comment:
This is nice to have but I think it ought to be a `Debugf` level log.

Reply:
👍